drake/solvers/trajectoryOptimization/PsmTraj.m

```diff
@@ -0,0 +1,445 @@
+      % add control inputs as bounding box constraints
+      if any(~isinf(plant.umin)) || any(~isinf(plant.umax))
+        control_limit = BoundingBoxConstraint(repmat(plant.umin,N,1),repmat(plant.umax,N,1));
+        obj = obj.addConstraint(control_limit,obj.u_inds(:));
+      end
+    end
+
+
+    function obj = addDynamicConstraints(obj)
+      N = obj.N;
+      nX = obj.plant.getNumStates();
+      nU = obj.plant.getNumInputs();
+
+      %only need the state and input value at each knot points
+      n_vars = (N)*(nX+nU)+1;
+      cnstr = FunctionHandleConstraint(zeros(N*(nX),1),zeros(N*(nX),1),n_vars,@obj.constraint_fun);
+      cnstr = setName(cnstr,'PSMcollocation');
+      % shared_data_index = obj.getNumSharedDataFunctions;
+      for i=1:obj.N,
+        obj = obj.addSharedDataFunction(@obj.dynamics_data,{obj.x_inds(:,i);obj.u_inds(:,i)});      
+      end 
+
+      all_x_inds=reshape(obj.x_inds(:,:),N*nX,1);
+      all_u_inds=reshape(obj.u_inds(:,:),N*nU,1);
+      obj = obj.addConstraint(cnstr,{obj.h_inds(1);all_x_inds;all_u_inds},[1:N]);
+    end
+
+
+    function [g,dg] = constraint_fun(obj,h1,x,u,varargin)
+    %%necessary parameters
+      N=obj.N;
+      nX = obj.plant.getNumStates();
+      nU = obj.plant.getNumInputs();
+      options=obj.options;
+      D=obj.D;
+      initial_h1=obj.tau(2)-obj.tau(1);
+      scale=h1/initial_h1;%comes from the time domain affine transformation
+
+      %%subscript p is for parameters from the psm method,f is for that from dynamics
+    %%from dynamics:
+      xdot_f=sparse(nX*N,1);
+      dxdh_f=sparse(nX*N,1);
+      dxdot_f=sparse(nX*N,nX*N);
+      udot_f=sparse(nX*N,N*nU);
+      dxdudot_f=sparse(N*nU,N*nX);
+      dxandudot_f=sparse(N*(nX+nU),N*(nX+nU)+1);
+    % from psm
+      xdot_p=sparse(nX*N,1);
+      u_p=sparse(nU*N,1);
+      dD_p=sparse(N*(nX+nU),N*(nX+nU));
+      dDxx_p=sparse(N*nX,N*nX);
+      dDxu_p=sparse(N*nX,N*nU);
+
+    %%f and df (the dynamics) evaluated at the knot points  
+      for k=1:N, %row indices
+        xdot_f((k-1)*nX+1:k*nX,:)= varargin{k}.xdot*scale; %%f(x,u), with scaling
+        dxdh_f((k-1)*nX+1:k*nX,:)= varargin{k}.xdot/initial_h1; %%df/dh 
+        dxdot_f((k-1)*nX+1:k*nX,(k-1)*nX+1:k*nX) = varargin{k}.dxdot(:,2:1+nX)*scale;%%df/dx, with scaling
+        udot_f((k-1)*nX+1:k*nX,(k-1)*nU+1:k*nU) = varargin{k}.dxdot(:,2+nX:end)*scale; %%df/du 
+        for j=1:N, %column indices
+          xdot_p((k-1)*(nX)+1:k*nX,:)=xdot_p((k-1)*(nX)+1:k*nX,:)+D(k,j)*x((j-1)*nX+1:j*nX);
+          dDxx_p((k-1)*(nX)+1:k*nX,(j-1)*(nX)+1:j*nX) = D(k,j)*eye(nX);
+          dDuu_p((k-1)*(nU)+1:k*nU,(j-1)*(nU)+1:j*nU) = D(k,j)*eye(nU);
+        end       
+      end 
+      
+    %%difference between the dynamic and collocation
+      g=[xdot_p]-[xdot_f];
+    %% gradient of such difference
+      dxandudot_f=[dxdh_f,dxdot_f,udot_f];
+      dD_p=[zeros(N*nX,1),dDxx_p,dDxu_p];
+      dg=sparse(dD_p)-sparse(dxandudot_f);      
+    end
```

Comment:
Cool! I did not know that `cnstr.setSparseStructure` trick. I'll definitely try that. 

I'll also try all the suggestions below, thanks for all of them!